src/uri.rs

```diff
@@ -1,11 +1,54 @@
 use std::borrow::Cow;
 use std::fmt::{Display, self};
-use std::str::FromStr;
+use std::ops::Deref;
+use std::str::{self, FromStr};
+use http::buf::{MemSlice, MemStr};
 use Url;
 use url::ParseError as UrlError;
 
 use Error;
 
+#[derive(PartialEq, Eq, Clone)]
+enum InternalUri {
+    Static(Cow<'static, str>),
```

Comment:
I'd rename this variant to `Cow`, since it is misleading when you have `InternalUri::Static(Cow::Owned("/foo".to_string()))`, because it looks like a static string, but it is owned.